README.md
[![Gem Version](https://badge.fury.io/rb/readthis.svg)](http://badge.fury.io/rb/readthis)
[![Build Status](https://travis-ci.org/sorentwo/readthis.svg?branch=master)](https://travis-ci.org/sorentwo/readthis)
[![Code Climate](https://codeclimate.com/github/sorentwo/readthis/badges/gpa.svg)](https://codeclimate.com/github/sorentwo/readthis)
[![Coverage Status](https://coveralls.io/repos/sorentwo/readthis/badge.svg?branch=master&service=github)](https://coveralls.io/github/sorentwo/readthis?branch=master)
[![Inline Docs](http://inch-ci.org/github/sorentwo/readthis.svg?branch=master)](http://inch-ci.org/github/sorentwo/readthis)

# Readthis

Readthis is a Redis backed cache client for Ruby. It is a drop in replacement
for any `ActiveSupport` compliant cache and can also be used for [session
storage](#session-storage). Above all Readthis emphasizes performance,
simplicity, and explicitness.

For new projects there isn't any reason to stick with Memcached. Redis is as
fast, if not faster in many scenarios, and is far more likely to be used
elsewhere in the stack. See [this blog post][hp-caching] for more details.

[hp-caching]: http://sorentwo.com/2015/07/20/high-performance-caching-with-readthis.html

## Footprint & Performance

See [Performance](PERFORMANCE.md)

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'readthis'
gem 'hiredis' # Highly recommended
```

## Usage

Use it the same way as any other [ActiveSupport::Cache::Store][store]. Within a
Rails environment config:

```ruby
config.cache_store = :readthis_store, {
  expires_in: 2.weeks.to_i,
  namespace: 'cache',
  redis: { url: ENV.fetch('REDIS_URL'), driver: :hiredis }
}
```

Otherwise you can use it anywhere, without any reliance on `ActiveSupport`:

```ruby
require 'readthis'

cache = Readthis::Cache.new(
  expires_in: 2.weeks.to_i,
  redis: { url: ENV.fetch('REDIS_URL') }
)
```

You can also specify `host`, `port`, `db` or any other valid Redis options. For
more details about connection options see in [redis gem documentation][redisrb]

[store]: http://api.rubyonrails.org/classes/ActiveSupport/Cache/Store.html
[redisrb]: https://github.com/redis/redis-rb#getting-started

### Instances & Databases

An isolated Redis instance that is only used for caching is ideal. Dedicated
instances have numerous benefits like: more predictable performance, avoiding
expires in favor of LRU, and tuning the persistence mechanism. See [Optimizing
Redis Usage for Caching][optimizing-usage] for more details.

At the very least, you'll want to use a specific database for caching. In the
event the database needs to be purged you can do so with a single `clear`
command, rather than finding all keys in a namespace and deleting them.
Appending a number between 0 and 15 will specify the redis database, which
defaults to `0`. For example, using database `2`:

```bash
REDIS_URL=redis://localhost:6379/2
```

[optimizing-usage]: http://sorentwo.com/2015/07/27/optimizing-redis-usage-for-caching.html

### Expiration

Be sure to use an integer value when setting expiration time. The default
representation of `ActiveSupport::Duration` values won't work when setting
expiration time, which will cause all keys to have `-1` as the TTL. Expiration
values are always cast as an integer on write. For example:

```ruby
Readthis::Cache.new(expires_in: 1.week) # don't do this
Readthis::Cache.new(expires_in: 1.week.to_i) # do this
```

By using the `refresh` option the TTL for keys can be refreshed automatically
every time the key is read. This is helpful for ensuring commonly hit keys are
kept cached, effectively making the cache a hybrid LRU.

```ruby
Readthis::Cache.new(refresh: true)
```

Be aware that `refresh` adds a slight overhead to all read operations, as they
are now all write operations as well.

### Compression

Compression can be enabled for all actions by passing the `compress` flag. By
default all values greater than 1024k will be compressed automatically. If there
is any content has not been stored with compression, or perhaps was compressed
but is beneath the compression threshold, it will be passed through as is. This
means it is safe to enable or change compression with an existing cache. There
will be a decoding performance penalty in this case, but it should be minor.

```ruby
config.cache_store = :readthis_store, {
  compress: true,
  compression_threshold: 2.kilobytes
}
```

### Serializing

Readthis uses Ruby's `Marshal` module for serializing all values by default.
This isn't always the fastest option, and depending on your use case it may be
desirable to use a faster but less flexible serializer.

By default Readthis knows about 3 different serializers:

* Marshal
* JSON
* Passthrough

If all cached data can safely be represented as a string then use the
pass-through serializer:

```ruby
Readthis::Cache.new(marshal: Readthis::Passthrough)
```

You can introduce up to four additional serializers by configuring `serializers`
on the Readthis module. For example, if you wanted to use the extremely fast Oj
library for JSON serialization:

```ruby
Readthis.serializers << Oj

# Freeze the serializers to ensure they aren't changed at runtime.
Readthis.serializers.freeze!

Readthis::Cache.new(marshal: Oj)
```

Be aware that the *order in which you add serializers matters*. Serializers are
sticky and a flag is stored with each cached value. If you subsequently go to
deserialize values and haven't configured the same serializers in the same order
your application will raise errors.

## Fault Tolerance

In some situations it is desirable to keep serving requests from disk or the
database if Redis crashes. This can be achieved with connection fault tolerance
by enabling it at the top level:

```ruby
Readthis.fault_tolerant = true
```

The default value is `false`, because while it may work for `fetch` operations,
it isn't compatible with other state-based commands like `increment`.

## Running Arbitrary Redis Commands

Readthis provides access to the underlying Redis connection pool, allowing you
to run arbitrary commands directly through the cache instance. For example, if
you wanted to expire a key manually using an instance of `Rails.cache`:

```ruby
Rails.cache.pool.with { |client| client.expire('foo-key', 60) }
```

## Differences From ActiveSupport::Cache

Readthis supports all of standard cache methods except for the following:

* `cleanup` - Redis does this with TTL or LRU already.
* `mute` and `silence!` - You must subscribe to the events `/cache*.active_support/`
  with `ActiveSupport::Notifications` to [log cache calls manually][notifications].

[notifications]: https://github.com/sorentwo/readthis/issues/22#issuecomment-142595938

Like other `ActiveSupport::Cache` implementations it is possible to cache `nil`
as a value. However, the fetch methods treat `nil` values as a cache miss and
re-generate/re-cache the value. Caching `nil` isn't recommended.

## Session Storage

By using [ActionDispatch::Session::CacheStore][cache-store] it's possible to
reuse `:readthis_store` or specify a new Readthis cache store for storing
sessions.

```ruby
Rails.application.config.session_store :cache_store
```

To specify a separate Readthis instance you can use the `:cache` option:

```ruby
Rails.application.config.session_store :cache_store,
  cache: Readthis::Cache.new,
  expire_after: 2.weeks.to_i
```

[cache-store]: http://api.rubyonrails.org/classes/ActionDispatch/Session/CacheStore.html

## Contributing

1. Fork it
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Add some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create new Pull Request


CHANGELOG.md
## v2.2.0 2018-01-14

- Added: `Cache#clear` now supports `async`, available on Redis 4.0+
- Added: Documentation for serializers and expanders
- Fixed: Preserve existing expiration value when incrementing or decrementing
  keys. There is also additional documentation about the non-atomicity of the
  increment and decrement methods. See [readthis#59][issue-59] for discussion.
- Fixed: Uniformly force key encodings to binary. Not all methods forced
  encoding of the key, which would cause incomatible encoding errors.
- Changed: Automatically duplicate frozen strings when expanding keys.

[issue-59]: https://github.com/sorentwo/readthis/issues/59

## v2.1.0 2017-10-07

- Changed: Allow any redis client version between 3.0 and 5.0

## v2.0.2 2016-12-12

- Fixed: Properly handle writing when the key and value are both UTF-8. Setting
  both the key and value to UTF-8 strings caused differing encodings to be sent
  to Redis. The key was UTF-8 but the value would be ASCII-8BIT, because it was
  compressed and then packed as 8bit integers. The workaround is to force the
  encoding of both strings to binary, something that will always be compatible.

  This makes Readthis actually usable with Non-English keys.

## v2.0.1 2016-07-27

- Fixed: The 2.0.0 release had a typo that rendered it useless.

## v2.0.0 2016-07-27

- Breaking: `nil` values are now rejected from `read_multi` output by default.
  Previously `nil` values were always preserved when fetching with `read_multi`.
  This differed from the `ActiveSupport::Cache` base behaviour and caused bugs
  in Rails 5.0. More details can be seen at [readthis#39][issue-49].

The old fetching behaviour, with `nil` intact, can be achieved by passing the
`retain_nils` flag:

```ruby
cache.read_multi('a', 'b', retain_nils: true)
```

This can also be set as a global option to retain backward compatibility:

```ruby
config.cache_store = :readthis_store, { retain_nils: true }
```

[issue-49]: https://github.com/sorentwo/readthis/issues/49

## v1.5.0 2016-07-18

- Added: `Cache#delete_matched` has been added in a resource efficient way that
  avoids using the "evil" `KEYS` command.

## v1.4.1 2016-07-05

- Fixed: Require lua files relative to the `Script` file itself. This fixes
  loading scripts from Rails or other applications.

## v1.4.0 2016-07-04

- Added: `Readthis::Script`, for dynamically loading and executing lua scripts.
- Added: Use custom `mexpire` to boost refresh performance with multiple keys.
  It is still slower than raw read performance, but much faster than naive
  `multi` expire (~1.5x faster).
- Changed: Generally tighten inline documentation.
- Fixed: Duplicate objects during `dump` operation when using the passthrough.
  When using `fetch` the original value was returned, including any encoding
  information that was prepended. From [readthis#44][issue-44] by @kagux
- Fixed: Only account for 3 bits when detecting serialized values, which fixes
  detection with 4 or more serializers configured. From [readthis#45][issue-45]
  by @kagux.
- Fixed: The max serializer guard would allow up to 8 serializers, which wasn't
  caught by an improper spec. From [readthis#46][issue-46] by @epilgrim.

[issue-44]: https://github.com/sorentwo/readthis/pull/44
[issue-45]: https://github.com/sorentwo/readthis/pull/45
[issue-46]: https://github.com/sorentwo/readthis/pull/46

## v1.3.0 2016-06-08

- Added: Key expiration refreshing on read. When `refresh: true` is set as an
  instance or method option all read operations will refresh the expiration.
- Fixed: Convert float `expires_in` values to the nearest valid integer.
  Typically any value less than 1 was rounded down to 0, which is an invalid
  expiration. Now a value of `0.1` will be rounded up to `1`, the lowest
  possible expiration.

## v1.2.1 2016-04-06

- Fixed: Splat arguments passed to `mget` within Readthis::Cache#read_multi.
  From [readthis#34][issue-34] submitted by @kyohei-shimada.

[issue-34]: https://github.com/sorentwo/readthis/pull/32

## v1.2.0 2015-12-16

- Added: Global connection fault tolerance. Any Redis connection error will
  raise be caught and a `nil` value will be returned instead. For `fetch`
  operations that means the block will yielded, if provided.
  [readthis#26][issue-26]

[issue-26]: https://github.com/sorentwo/readthis/issues/26

## v1.1.0 2015-12-07

- Fixed: Stop overwriting specific options with the default options. [issue-28].
  Discovered and fixed by @tobinibot.
- Fixed: Handle the case when `nil` is explicitly passed as options to `fetch`.
- Changed: All errors now extend from a base `ReadthisError`.

[issue-28]: https://github.com/sorentwo/readthis/issues/28

## v1.0.0 2015-10-05

- Changed: Remove internal `Notifications` module as part of instrumentation
  cleanup.

## v1.0.0-rc.1 2015-09-27

- Fixed: Custom serializers would be encoded correcty, but would be ignored when
  the value was read back out, leaving the encoding flags prefixed to the value.
- Changed: There are no longer direct accessors for `namespace` and
  `expires_in`. Instead, the exposed `options` hash is used as the fallback for
  all operations. For `ActiveSupport` compliance options must be exposed, so this
  prevents configuration drift after initialization. [readthis#21][issue-21]
- Added: More helpful errors are raised when attempting to use a serializer that
  hasn't been configured. [readthis#22][issue-22].

[issue-21]: https://github.com/sorentwo/readthis/issues/21
[issue-22]: https://github.com/sorentwo/readthis/issues/22

## v1.0.0-beta 2015-09-18

- Breaking: This change is necessary for the consistency and portability of
  values going forward. All entities are now written with a set of option flags
  as the initial byte. This flag is later used to determine whether the entity
  was compressed and what was used to marshal it. There are a number of
  advantages to this approach, consistency and reliability being the most

lib/readthis/cache.rb
# frozen_string_literal: true

require 'readthis/entity'
require 'readthis/expanders'
require 'readthis/passthrough'
require 'readthis/scripts'
require 'redis'
require 'connection_pool'

module Readthis
  # Readthis is a Redis backed cache client. It is a drop in replacement for
  # any `ActiveSupport` compliant cache Above all Readthis emphasizes
  # performance, simplicity, and explicitness.
  class Cache
    attr_reader :entity, :notifications, :options, :pool, :scripts

    # Provide a class level lookup of the proper notifications module.
    # Instrumention is expected to occur within applications that have
    # ActiveSupport::Notifications available, but needs to work even when it
    # isn't.
    def self.notifications
      ActiveSupport::Notifications if defined?(ActiveSupport::Notifications)
    end

    # Creates a new Readthis::Cache object with the given options.
    #
    # @option options [Hash] :redis Options that will be passed to the redis
    #   connection
    # @option options [Boolean] :compress (false) Enable or disable automatic
    #   compression
    # @option options [Number] :compression_threshold (8k) Minimum string size
    #   for compression
    # @option options [Number] :expires_in The number of seconds until an entry
    #   expires
    # @option options [Boolean] :refresh (false) Automatically refresh key
    #   expiration
    # @option options [Boolean] :retain_nils (false) Whether nil values should
    #   be included in read_multi output
    # @option options [Module] :marshal (Marshal) Module that responds to
    #   `dump` and `load`
    # @option options [String] :namespace Prefix used to namespace entries
    # @option options [Number] :pool_size (5) The number of threads in the pool
    # @option options [Number] :pool_timeout (5) How long before a thread times
    #   out
    #
    # @example Create a new cache instance
    #
    #   Readthis::Cache.new(namespace: 'cache',
    #                       redis: { url: 'redis://localhost:6379/0' })
    #
    # @example Create a compressed cache instance
    #
    #   Readthis::Cache.new(compress: true, compression_threshold: 2048)
    #
    def initialize(options = {})
      @options = options

      @entity = Readthis::Entity.new(
        marshal: options.fetch(:marshal, Marshal),
        compress: options.fetch(:compress, false),
        threshold: options.fetch(:compression_threshold, 1024)
      )

      @pool = ConnectionPool.new(pool_options(options)) do
        Redis.new(options.fetch(:redis, {}))
      end

      @scripts = Readthis::Scripts.new
    end

    # Fetches data from the cache, using the given key. If there is data in
    # the cache with the given key, then that data is returned. Otherwise, nil
    # is returned.
    #
    # @param [String] key Key for lookup
    # @param [Hash] options Optional overrides
    #
    # @example
    #
    #   cache.read('missing') # => nil
    #   cache.read('matched') # => 'some value'
    #
    def read(key, options = {})
      options = merged_options(options)

      invoke(:read, key) do |store|
        key = namespaced_key(key, options)

        refresh_entity(key, store, options)

        entity.load(store.get(key))
      end
    end

    # Writes data to the cache using the given key. Will overwrite whatever
    # value is already stored at that key.
    #
    # @param [String] key Key for lookup
    # @param [Hash] options Optional overrides
    #
    # @example
    #
    #   cache.write('some-key', 'a bunch of text')                     # => 'OK'
    #   cache.write('some-key', 'short lived', expires_in: 60)         # => 'OK'
    #   cache.write('some-key', 'lives elsehwere', namespace: 'cache') # => 'OK'
    #
    def write(key, value, options = {})
      options = merged_options(options)

      invoke(:write, key) do |store|
        write_entity(key, value, store, options)
      end
    end

    # Delete the value stored at the specified key. Returns `true` if
    # anything was deleted, `false` otherwise.
    #
    # @param [String] key The key for lookup
    # @param [Hash] options Optional overrides
    #
    # @example
    #
    #   cache.delete('existing-key') # => true
    #   cache.delete('random-key')   # => false
    #
    def delete(key, options = {})
      namespaced = namespaced_key(key, merged_options(options))

      invoke(:delete, key) do |store|
        store.del(namespaced) > 0
      end
    end

    # Delete all values that match a given pattern. The pattern must be defined
    # using Redis compliant globs. The following examples are borrowed from the
    # `KEYS` documentation:
    #
    # * `h?llo` matches hello, hallo and hxllo
    # * `h*llo` matches hllo and heeeello
    # * `h[ae]llo` matches hello and hallo, but not hillo
    # * `h[^e]llo` matches hallo, hbllo, ... but not hello
    # * `h[a-b]llo` matches hallo and hbllo
    #
    # Note that `delete_matched` does *not* use the `KEYS` command, making it
    # safe for use in production.
    #
    # @param [String] pattern The glob pattern for matching keys
    # @option [String] :namespace Prepend a namespace to the pattern
    # @option [Number] :count Configure the number of keys deleted at once
    #
    # @example Delete all 'cat' keys
    #
    #   cache.delete_matched('*cats') #=> 47
    #   cache.delete_matched('*dogs') #=> 0
    #
    def delete_matched(pattern, options = {})
      namespaced = namespaced_key(pattern, merged_options(options))

      invoke(:delete, pattern) do |store|
        cursor = nil
        count = options.fetch(:count, 1000)
        deleted = 0

        until cursor == '0'
          cursor, matched = store.scan(cursor || 0, match: namespaced, count: count)

          if matched.any?
            store.del(*matched)
            deleted += matched.length
          end
        end

        deleted
      end
    end

    # Fetches data from the cache, using the given key. If there is data in the
    # cache with the given key, then that data is returned.
    #
    # If there is no such data in the cache (a cache miss), then `nil` will be
    # returned. However, if a block has been passed, that block will be passed
    # the key and executed in the event of a cache miss. The return value of
    # the block will be written to the cache under the given cache key, and
    # that return value will be returned.
    #
    # @param [String] key Key for lookup
    # @param options [Hash] Optional overrides
    # @option options [Boolean] :force Force a cache miss
    # @yield [String] Gives a missing key to the block, which is used to
    #   generate the missing value
    #
    # @example Typical
    #
    #   cache.write('today', 'Monday')
    #   cache.fetch('today') # => "Monday"
    #   cache.fetch('city')  # => nil
    #
    # @example With a block
    #
    #   cache.fetch('city') do
    #     'Duckburgh'
    #   end
    #
    #   cache.fetch('city') # => "Duckburgh"
    #
    # @example Cache Miss
    #
    #   cache.write('today', 'Monday')
    #   cache.fetch('today', force: true) # => nil
    #
    def fetch(key, options = {})
      options ||= {}
      value = read(key, options) unless options[:force]

      if value.nil? && block_given?
        value = yield(key)
        write(key, value, options)
      end

      value
    end

    # Increment a key in the store.
    #
    # If the key doesn't exist it will be initialized at 0. If the key exists
    # but it isn't a Fixnum it will be coerced to 0.
    #
    # Note that this method does *not* use Redis' native `incr` or `incrby`
    # commands. Those commands only work with number-like strings, and are
    # incompatible with the encoded values Readthis writes to the store. The
    # behavior of `incrby` is preserved as much as possible, but incrementing
    # is not an atomic action. If multiple clients are incrementing the same
    # key there will be a "last write wins" race condition, causing incorrect
    # counts.
    #
    # If you absolutely require correct counts it is better to use the Redis
    # client directly.
    #
    # @param [String] key Key for lookup
    # @param [Fixnum] amount Value to increment by
    # @param [Hash] options Optional overrides
    #
    # @example
    #
    #   cache.increment('counter') # => 1
    #   cache.increment('counter', 2) # => 3
    #
    def increment(key, amount = 1, options = {})
      invoke(:increment, key) do |store|
        alter(store, key, amount, options)
      end
    end

    # Decrement a key in the store.
    #
    # If the key doesn't exist it will be initialized at 0. If the key exists
    # but it isn't a Fixnum it will be coerced to 0. Like `increment`, this
    # does not make use of the native `decr` or `decrby` commands.
    #
    # @param [String] key Key for lookup
    # @param [Fixnum] amount Value to decrement by
    # @param [Hash] options Optional overrides
    #
    # @example
    #
    #   cache.write('counter', 20) # => 20
    #   cache.decrement('counter') # => 19
    #   cache.decrement('counter', 2) # => 17
    #
    def decrement(key, amount = 1, options = {})
      invoke(:decrement, key) do |store|
        alter(store, key, -amount, options)
      end
    end

    # Efficiently read multiple values at once from the cache. Options can be
    # passed in the last argument.
    #
    # @overload read_multi(keys)
    #   Return all values for the given keys.
    #   @param [String] One or more keys to fetch
    #   @param [Hash] options Configuration to override
    #
    # @return [Hash] A hash mapping keys to the values found.
    #
    # @example
    #
    #   cache.read_multi('a', 'b') # => { 'a' => 1 }
    #   cache.read_multi('a', 'b', retain_nils: true) # => { 'a' => 1, 'b' => nil }
    #
    def read_multi(*keys)
      options = merged_options(extract_options!(keys))
      mapping = keys.map { |key| namespaced_key(key, options) }

      return {} if keys.empty?

      invoke(:read_multi, keys) do |store|
        values = store.mget(*mapping).map { |value| entity.load(value) }

        refresh_entity(mapping, store, options)

        zipped_results(keys, values, options)
      end
    end

    # Write multiple key value pairs simultaneously. This is an atomic
    # operation that will always succeed and will overwrite existing
    # values.
    #
    # This is a non-standard, but useful, cache method.
    #
    # @param [Hash] hash Key value hash to write
    # @param [Hash] options Optional overrides
    #
    # @example
    #
    #   cache.write_multi({ 'a' => 1, 'b' => 2 }) # => true
    #
    def write_multi(hash, options = {})
      options = merged_options(options)

      invoke(:write_multi, hash.keys) do |store|
        store.multi do
          hash.each { |key, value| write_entity(key, value, store, options) }
        end
      end
    end

    # Fetches multiple keys from the cache using a single call to the server
    # and filling in any cache misses. All read and write operations are
    # executed atomically.
    #
    # @overload fetch_multi(keys)
    #   Return all values for the given keys, applying the block to the key
    #   when a value is missing.
    #   @param [String] One or more keys to fetch
    #
    # @example
    #
    #   cache.fetch_multi('alpha', 'beta') do |key|
    #     "#{key}-was-missing"
    #   end
    #
    #   cache.fetch_multi('a', 'b', expires_in: 60) do |key|
    #     key * 2
    #   end
    #
    def fetch_multi(*keys)
      options = extract_options!(keys).merge(retain_nils: true)
      results = read_multi(*keys, options)
      missing = {}

      invoke(:fetch_multi, keys) do |_store|
        results.each do |key, value|
          next unless value.nil?

          value = yield(key)
          missing[key] = value
          results[key] = value
        end
      end

      write_multi(missing, options) if missing.any?

      results
    end

    # Returns `true` if the cache contains an entry for the given key.
    #
    # @param [String] key Key for lookup
    # @param [Hash] options Optional overrides
    #
    # @example
    #
    #   cache.exist?('some-key') # => false
    #   cache.exist?('some-key', namespace: 'cache') # => true
    #
    def exist?(key, options = {})
      invoke(:exist?, key) do |store|
        store.exists(namespaced_key(key, merged_options(options)))
      end
    end

    # Clear the entire cache by flushing the current database.
    #
    # This flushes everything in the current database, with no globbing
    # applied. Data in other numbered databases will be preserved.
    #
    # @option options [Hash] :async Flush the database asynchronously, only
    #   supported in Redis 4.0+
    #
    # @example
    #
    #   cache.clear #=> 'OK'
    #   cache.clear(async: true) #=> 'OK'
    #
    def clear(options = {})
      invoke(:clear, '*') do |store|
        if options[:async]
          store.flushdb(async: true)
        else
          store.flushdb
        end
      end
    end

    protected

    def refresh_entity(keys, store, options)
      return unless options[:refresh] && options[:expires_in]

      expiration = coerce_expiration(options[:expires_in])

      scripts.run('mexpire', store, keys, expiration)
    end

    def write_entity(key, value, store, options)
      namespaced = namespaced_key(key, options)
      dumped = entity.dump(value, options)

      if (expiration = options[:expires_in])
        store.setex(namespaced, coerce_expiration(expiration), dumped)
      else
        store.set(namespaced, dumped)
      end
    end

    private

    def alter(store, key, amount, options)
      options = merged_options(options)
      namespaced = namespaced_key(key, options)

      loaded = entity.load(store.get(namespaced))
      change = loaded.to_i + amount
      dumped = entity.dump(change, options)
      expiration = fallback_expiration(store, namespaced, options)

      if expiration
        store.setex(namespaced, coerce_expiration(expiration), dumped)
      else
        store.set(namespaced, dumped)
      end

      change
    end

    def fallback_expiration(store, key, options)
      options.fetch(:expires_in) do
        ttl = store.ttl(key)

        ttl > 0 ? ttl : nil
      end
    end

    def coerce_expiration(expires_in)
      Float(expires_in).ceil
    end

    def instrument(name, key)
      if self.class.notifications
        name = "cache_#{name}.active_support"
        payload = { key: key, name: name }

        self.class.notifications.instrument(name, payload) { yield(payload) }
      else
        yield
      end
    end

    def invoke(operation, key, &block)
      instrument(operation, key) do
        pool.with(&block)
      end
    rescue Redis::BaseError => error
      raise error unless Readthis.fault_tolerant?
    end

    def extract_options!(array)
      array.last.is_a?(Hash) ? array.pop : {}
    end

    def merged_options(options)
      @options.merge(options || {})
    end

    def zipped_results(keys, values, options)
      zipped = keys.zip(values)

      zipped.select! { |(_, value)| value } unless options[:retain_nils]

      zipped.to_h
    end

    def pool_options(options)
      { size: options.fetch(:pool_size, 5),
        timeout: options.fetch(:pool_timeout, 5) }
    end

    def namespaced_key(key, options)
      Readthis::Expanders.namespace_key(key, options[:namespace])
    end
  end
end
